### PR TITLE
Transform all GPU symbols to become unique

### PR DESCRIPTION
**Pull request by wdeconinck**
We want to achieve unique symbols in all 4 ectrans flavours (combinations of  gpu/cpu with sp/dp).
This paves the way to create a future ectrans API that can dispatch to these 4 libraries at runtime.

For the CPU code (cpu + sp/dp) the unique symbols were already achieved in #61. 
The recently merged GPU code was not yet done, and this PR tackles this and follows the same approach.

The GPU symbols now have a suffix `_gpu_sp` for single precision and `_gpu_dp` for double precision
The library with these is `ectrans_gpu_sp` and `ectrans_gpu_dp` respectively and does not rely on any `parkind1` module.
It is therefore now possible to link to all of these together.
 - `ectrans_sp` (cpu)
 - `ectrans_dp` (cpu)
 - `ectrans_gpu_sp`
 - `ectrans_gpu_dp`

For backward compatibility there are interface libraries named as before `trans_gpu_sp` and `trans_gpu_dp` which give access to the previous API without the suffixes via preprocessor definitions. This is what currently allows `ectrans-benchmark.F90` to be generic. Naturally you cannot link to these simultaneously, which is why there are the 4 separately compiled flavours of `ectrans-benchmark` necessary.

```mermaid
graph BT;
    ectrans_gpu_sp-->ectrans_gpu_common;
    ectrans_gpu_dp-->ectrans_gpu_common;
    ectrans_gpu_common-->ectrans_common;
    ectrans_sp-->ectrans_common;
    ectrans_dp-->ectrans_common;
```
The above graph shows interdependence between libraries where the `_common` libraries contain only precision independent code.

- `ectrans_common` library contains domain decomposition routines as well as `setup_trans0`! A new test has been added to test `setup_trans0` which only links with `ectrans_common`.
- `ectrans_gpu_common` contains GPU specific routines (e.g. ACC, allocators, hip/cuda abstractions)

### CODE REVIEW

**Comment by samhatfield** [comment]
Nice work! I'll have a read through. Might take some time...

**Comment by samhatfield** [comment]
external/ini_spec_dist.F90 could also be moved to common. The imported `JPRB` in this file is not actually used, and the subroutine it calls `SUWAVEDI` is already in common.

File: src/trans/gpu/external/setup_trans0.F90
Comment (samhatfield) [inline src/trans/gpu/external/setup_trans0.F90:258]: KOPT_MEMORY_TR does nothing for the GPU version, hence this warning. Without this block in the new common SETUP_TRANS0, when the user specifies KOPT_MEMORY_TR and they use the GPU version they will not be told that the option is not actually doing anything. Do we want that behaviour?

File: src/trans/common/external/setup_trans0.F90
Comment (wdeconinck) [reply src/trans/common/external/setup_trans0.F90:148]: With the current approach setup_trans0 is in a common library which does not have knowledge which backend is being used. These specifics could go into a more specific 'setup_trans' though.

File: src/trans/gpu/external/setup_trans0.F90
Comment (wdeconinck) [reply src/trans/gpu/external/setup_trans0.F90:258]: The common setup_trans0 has no knowledge of what these values should be, and could not warn.
Perhaps the GPU-specific setup_trans could later query the value of NOPT_MEMORY_TR and choose how to respond, but I would choose anyway to ignore it if it's an unused option ¯\_(ツ)_/¯ .
Perhaps in the future also the gpu code may be adapted to also use the "stack" allocations for CPU specific arrays.

**Comment by wdeconinck** [comment]
> external/ini_spec_dist.F90 could also be moved to common. The imported `JPRB` in this file is not actually used, and the subroutine it calls `SUWAVEDI` is already in common.

Now done

File: src/trans/common/external/setup_trans0.F90
Comment (wdeconinck) [reply src/trans/common/external/setup_trans0.F90:148]: I have now added this diagnostic print as well when `NPRINTLEV>0` to `setup_trans` which is compiled within the 4 sublibraries `ectrans_(gpu_){sp,dp}`. It now prints the used precision as well.

**Comment by wdeconinck** [comment]
Changes to HICFFT will have to be revisited and will probably be much simpler after first having merged #143

File: src/trans/common/external/setup_trans0.F90
Comment (samhatfield) [reply src/trans/common/external/setup_trans0.F90:148]: That's great, thanks.

**Comment by wdeconinck** [comment]
> Changes to HICFFT will have to be revisited and will probably be much simpler after first having merged #143

This is now done. I deleted more unused HICFFT code in develop branch and then rebased this branch.

File: src/trans/gpu/CMakeLists.txt
Comment (lukasm91) [inline src/trans/gpu/CMakeLists.txt:132]: make sure to properly apply https://github.com/ecmwf-ifs/ectrans/commit/51cd73c0e959f3ead89a2033fa6f3553609e885a

**Comment by samhatfield** [comment]
I think the following files can be added to the common library:

- pre_suleg_mod.F90
- setup_dims_mod.F90
- setup_dims_mod.F90
- setup_geom_mod.F90
- shuffle_mod.F90
- sutrle_mod.F90

**Review by samhatfield — CHANGES_REQUESTED** [review]
Very nice work here. I'm happy to see ~5,000 lines deleted. Happy with the merge, subject to the comments above.

**Comment by wdeconinck** [comment]
> Here's a test program which attempts to call `SETUP_TRANS` for CPU and GPU:
> 
> ```fortran
> PROGRAM ECTRANS_DEMONSTRATOR
> 
> USE EC_PARKIND, ONLY: JPIM
> 
> IMPLICIT NONE
> 
> #include "setup_trans0.h"
> #include "setup_trans_sp.h"
> #include "setup_trans_gpu_sp.h"
> 
> INTEGER(JPIM), PARAMETER :: TRUNC = 79
> 
> CALL SETUP_TRANS0(LDMPOFF=.TRUE., KMAX_RESOL=2, KPRINTLEV=2)
> 
> CALL SETUP_TRANS_SP(KSMAX=TRUNC, KDGL=2 * (TRUNC + 1)) ! CPU
> CALL SETUP_TRANS_GPU_SP(KSMAX=TRUNC, KDGL=2 * (TRUNC + 1)) ! GPU
> 
> END PROGRAM ECTRANS_DEMONSTRATOR
> ```
> 
> with Makefile
> 
> ```makefile
> FC = nvfortran
> 
> ECTRANS_BUILD = /ec/res4/hpcperm/nash/ectrans/build
> FIAT_BUILD = /ec/res4/hpcperm/nash/fiat/build
> 
> ectrans_demonstrator: ectrans_demonstrator.o
>     $(FC) $^ -o ectrans_demonstrator -Wl,-rpath,$(ECTRANS_BUILD)/lib \
>     -Wl,-rpath,$(FIAT_BUILD)/lib $(ECTRANS_BUILD)/lib/libectrans_common.so \
>     $(ECTRANS_BUILD)/lib/libectrans_sp.so $(ECTRANS_BUILD)/lib/libectrans_gpu_sp.so \
>     $(FIAT_BUILD)/lib/libfiat.so
> 
> ectrans_demonstrator.o: ectrans_demonstrator.F90
> 
> %.o: %.F90
>     $(FC) -I$(ECTRANS_BUILD)/include/ectrans \
>     -module $(FIAT_BUILD)/module/fiat -c $< -o $@
> ```
> 
> I don't think we're quite ready for this kind of program. When I run this I get a segfault in `SETUP_TRANS_GPU_SP` on [this line](https://github.com/wdeconinck/ectrans/blob/feature/unique_symbols_gpu/src/trans/gpu/external/setup_trans.F90#L272).
> 
> Why this line and not the ones above? `G` and `D` are defined in `TPM_GEOMETRY` and `TPM_DISTR`, respectively, which are both in the common library. They are pointers to arrays which are allocated and visible in this scope. `S` however is defined in `TPM_FLT` which is CPU/GPU-specific. It's a pointer to an element of `FLT_RESOL`. `FLT_RESOL` in `TPM_FLT_SP` is allocated (during the first call to `SETUP_TRANS`) but `FLT_RESOL` in `TPM_FLT_GPU_SP` is still unallocated at this point. Hence the segfault.
> 
> We can perhaps solve this one by moving `TPM_FLT` to common, but I wonder if there are other potential conflicts?

I think we're still a long way to be able to create multiple handles that belong to different backends, even though that is our ultimate aim.
At least it should be possible to have a runtime switch to call either one of them, and stick to the same backend for the rest of the run.

File: src/trans/gpu/external/sugawc.F90
Comment (wdeconinck) [reply src/trans/gpu/external/sugawc.F90:1]: Yes it did, but I moved it recently to "src/trans/cpu/maybe_unused" directory and it is not compiled.
If nobody is complaining this is dead code and can be removed in near future, otherwise easy to revive.
The gpu/external version is an exact copy of this. If it needs to be revived it can likely go in common.

File: src/trans/gpu/CMakeLists.txt
Comment (wdeconinck) [reply src/trans/gpu/CMakeLists.txt:132]: I will rebase and take extra care.

**Comment by wdeconinck** [comment]
> I think the following files can be added to the common library:
> 
> * pre_suleg_mod.F90
> * setup_dims_mod.F90
> * setup_geom_mod.F90
> * shuffle_mod.F90
> * sutrle_mod.F90

From all the above, only shuffle_mod.F90 can be moved to common (now done). The others are using either `TPM_FIELDS` or "TPM_FLT" which contain JPRB data.
Refactoring this code would be great, and possibly a lot can be made common.
e.g. computing and storing the Legendre coefficients only in double precision, and be reused by multiple backends of same resolution too!

File: tests/trans/test_setup_trans0.F90
Comment (wdeconinck) [reply tests/trans/test_setup_trans0.F90]: Done

File: src/trans/gpu/CMakeLists.txt
Comment (wdeconinck) [reply src/trans/gpu/CMakeLists.txt]: Done

**Comment by samhatfield** [comment]
> > I think the following files can be added to the common library:
> > 
> > * pre_suleg_mod.F90
> > * setup_dims_mod.F90
> > * setup_geom_mod.F90
> > * shuffle_mod.F90
> > * sutrle_mod.F90
> 
> From all the above, only shuffle_mod.F90 can be moved to common (now done). The others are using either `TPM_FIELDS` or `TPM_FLT` which contain `JPRB` data. Refactoring this code would be great, and possibly a lot can be made common. e.g. computing and storing the Legendre coefficients only in double precision, and be reused by multiple backends of same resolution too!

Ah yes I see. I will think about how we can move more TPM_ modules to common.

**Review by samhatfield — APPROVED** [review]

**Comment by wdeconinck** [comment]
Now rebased and amended relevant commits with the fixups.
I will merge when @marsdeno agrees.

**Pull request merged**
